src/services/services.ts

```diff
@@ -4464,6 +4510,477 @@ namespace ts {
+                    }
+                }
+
+                return deduplicate(nonRelativeModules);
+            }
+
+            function getTripleSlashReferenceCompletion(sourceFile: SourceFile, position: number): ImportCompletionInfo {
+                const token = getTokenAtPosition(sourceFile, position);
+                if (!token) {
+                    return undefined;
+                }
+                const commentRanges: CommentRange[] = getLeadingCommentRanges(sourceFile.text, token.pos);
+
+                if (!commentRanges || !commentRanges.length) {
+                    return undefined;
+                }
+
+                const range = forEach(commentRanges, commentRange => position >= commentRange.pos && position <= commentRange.end && commentRange);
+
+                if (!range) {
+                    return undefined;
+                }
+
+                const text = sourceFile.text.substr(range.pos, position - range.pos);
+
+                const match = tripleSlashDirectiveFragmentRegex.exec(text);
+                if (match) {
+                    const prefix = match[1];
+                    const kind = match[2];
+                    const toComplete = match[3];
+
+                    const scriptPath = getDirectoryPath(sourceFile.path);
+                    if (kind === "path") {
+                        // Give completions for a relative path
+                        const textSpan: TextSpan = getDirectoryFragmentTextSpan(toComplete, range.pos + prefix.length);
+                        return {
+                            entries: getCompletionEntriesForDirectoryFragment(toComplete, scriptPath, getSupportedExtensions(program.getCompilerOptions()), /*includeExtensions*/true, sourceFile.path),
+                            textSpan
+                        };
+                    }
+                    else {
+                        // Give completions based on the typings available
+                        const textSpan: TextSpan = { start: range.pos + prefix.length, length: match[0].length - prefix.length };
+                        return {
+                            entries: getCompletionEntriesFromTypings(host, program.getCompilerOptions(), scriptPath),
+                            textSpan
+                        };
+                    }
+                }
+
+                return undefined;
+            }
+
+            function getCompletionEntriesFromTypings(host: LanguageServiceHost, options: CompilerOptions, scriptPath: string, result: ImportCompletionEntry[] = []): ImportCompletionEntry[] {
+                // Check for typings specified in compiler options
+                if (options.types) {
+                    for (const moduleName of options.types) {
+                        result.push(createCompletionEntryForModule(moduleName, ScriptElementKind.externalModuleName));
+                    }
+                }
+                else if (host.getDirectories && options.typeRoots) {

```

Comment:
get the effective typeRoots here, see `getEffectiveTypeRoots` in program. we can export this function and use it here as well.